src/appleseed/renderer/kernel/lighting/lighttree.cpp

```diff
@@ -0,0 +1,333 @@
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "lighttree.h"
+
+// appleseed.renderer headers.
+#include "renderer/global/globallogger.h"
+#include "renderer/global/globaltypes.h"
+
+// appleseed.foundation headers.
+#include "foundation/image/colorspace.h"
+#include "foundation/math/distance.h"
+#include "foundation/math/permutation.h"
+#include "foundation/platform/timers.h"
+#include "foundation/utility/foreach.h"
+#include "foundation/utility/vpythonfile.h"
+
+namespace renderer
+{
+
+//
+// LightTree class implementation.
+//
+
+LightTree::LightTree()
+  :m_tree_depth(0)
+{
+    m_built = false;
+}
+
+LightTree::~LightTree()
+{
+    RENDERER_LOG_INFO("Deleting the light tree...");
+    for(foundation::const_each<LightSourcePointerVector> i = m_light_sources; i; ++i)
+    {
+        delete *i;
+    }
+}
+
+bool LightTree::is_built() const
+{
+    return m_built;
+}
+
+void LightTree::build(
+        const std::vector<NonPhysicalLightInfo>     non_physical_lights)
+{
+    foundation::Statistics statistics;
+    AABBVector light_bboxes;
+
+    size_t light_index = 0;
+
+    // Collect all possible light sources into one vector
+    for (size_t i = 0; i < non_physical_lights.size(); ++i)
+    {
+        LightSource* light_source = new NonPhysicalLightSource(&non_physical_lights[i]);
+        foundation::AABB3d bbox = light_source->get_bbox();
+        m_light_sources.push_back(light_source);
+        light_bboxes.push_back(bbox);
+        m_items.push_back(
+            Item(
+                bbox,
+                light_index++,
+                i));
+    }
+
+    // Create the partitioner.
+    typedef foundation::bvh::MiddlePartitioner<AABBVector> Partitioner;
+    Partitioner partitioner(light_bboxes);
+
+    // Build the light tree.
+    typedef foundation::bvh::Builder<LightTree, Partitioner> Builder;
+    Builder builder;
+    builder.build<foundation::DefaultWallclockTimer>(*this, partitioner, m_items.size(), 1);
+    statistics.insert_time("build time", builder.get_build_time());
+
+    // Reorder m_items vector to match the ordering in the LightTree.
+    if (!m_items.empty())
+    {
+        m_built = true;
+
+        const std::vector<size_t>& ordering = partitioner.get_item_ordering();
+        assert(m_items.size() == ordering.size());
+
+        // Reorder the items according to the tree ordering.
+        ItemVector temp_lights(ordering.size());
+        foundation::small_item_reorder(
+            &m_items[0],
+            &temp_lights[0],
+            &ordering[0],
+            ordering.size());
+
+        // Set total luminance and level for each node of the LightTree.
+        recursive_node_update(0, 0);
+    }
+    
+    // Print light tree statistics.
+    RENDERER_LOG_INFO("%s",
+        foundation::StatisticsVector::make(
+            "light tree statistics",
+            statistics).to_string().c_str());
+
+    RENDERER_LOG_INFO("Number of light sources: %zu", m_light_sources.size());
```

Comment:
For some reason our log messages are in all lower case. For the sake of consistency, use lower case only :)